### PR TITLE
5176 sad path

### PR DESCRIPTION
**Pull request by sfkamath**
This branch handles sad path/edge cases of a user giving an incorrect password for login and different levels of assurance being selected

### CODE REVIEW

File: src/main/java/uk/gov/ida/eidas/bridge/helpers/requestToEidas/EidasAuthnRequestGenerator.java [hunk 1]
```diff
@@ -65,7 +66,17 @@ public AuthnRequest generateAuthnRequest(String authnReqeustId, String destinati
         eidasAuthnRequest.setNameIDPolicy(nameIdPolicy);
 
         RequestedAuthnContext requestedAuthnContext = openSamlXmlObjectFactory.createRequestedAuthnContext(AuthnContextComparisonTypeEnumeration.MINIMUM);
-        requestedAuthnContext.getAuthnContextClassRefs().add(openSamlXmlObjectFactory.createAuthnContextClassReference(LevelOfAssurance.SUBSTANTIAL.toString()));
+        String levelOfAssuranceRequested;
```
Comment (chrisholmes) [inline src/main/java/uk/gov/ida/eidas/bridge/helpers/requestToEidas/EidasAuthnRequestGenerator.java]: can we extract this somewhere?

File: src/main/java/uk/gov/ida/eidas/bridge/resources/VerifyAuthnRequestResource.java [hunk 2]
```diff
@@ -54,9 +59,23 @@ public Response receiveAuthnRequest(
 
         String authnRequestId = authnRequest.getID();
 
+        RequestedAuthnContext requestedAuthnContext = authnRequest.getRequestedAuthnContext();
+        List<AuthnContextClassRef> authnContextClassRefs = requestedAuthnContext.getAuthnContextClassRefs();
+        AuthnContext lowestAuthnContext = null;
+        AuthnContextFactory authnContextFactory = new AuthnContextFactory();
+        for(AuthnContextClassRef authnContextClassRef : authnContextClassRefs) {
```
Comment (richardTowers) [inline src/main/java/uk/gov/ida/eidas/bridge/resources/VerifyAuthnRequestResource.java]: At the moment when there's 0 authnContextClassRefs lowestAuthnContext will end up as null. Should probably throw an error in that case.

Also something like
```
lowest = authnContextClassRefs.stream()
    .map(x -> authnContextFactory.authnContextForLevelOfAssurance(x.getAuthnContextClassRef())
    .min(AuthContext::compare)
    .orElseThrow(new SomeException())
```

(untested)

Would be a more readable way of doing this in my opinion.

File: src/main/java/uk/gov/ida/eidas/bridge/helpers/requestToEidas/EidasAuthnRequestGenerator.java (hunk 1 above)
Comment (sfkamath) [inline src/main/java/uk/gov/ida/eidas/bridge/helpers/requestToEidas/EidasAuthnRequestGenerator.java]: done, extracted to a class constant 
or did you mean to an environment variable?

File: src/main/java/uk/gov/ida/eidas/bridge/resources/VerifyAuthnRequestResource.java [hunk 3]
```diff
@@ -54,9 +59,23 @@ public Response receiveAuthnRequest(
...
+            AuthnContext authnContext = authnContextFactory.authnContextForLevelOfAssurance(authnContextClassRef.getAuthnContextClassRef());
+            //TODO : check if the comparator is setup right !
```
Comment (sfkamath) [inline src/main/java/uk/gov/ida/eidas/bridge/resources/VerifyAuthnRequestResource.java]: removed, checked the compareTo javadoc before committing but left the TODO

File: src/main/java/uk/gov/ida/eidas/bridge/helpers/responseFromEidas/ResponseHandler.java [hunk 4]
```diff
@@ -39,15 +45,26 @@ public EidasSamlResponse handleResponse(String base64EncodedResponse, String exp
...
+            throw new SecurityException(RESPONSE_ISSUER_FAILURE_MESSAGE_START + " (" + actualIssuer + ") didn't match expected issuer (" + eidasEntityId + ")");
         }
 
         ValidatedResponse validatedResponse = samlResponseSignatureValidator.validate(response, IDPSSODescriptor.DEFAULT_ELEMENT_NAME);
+        Status status = validatedResponse.getStatus();
+        if(status != null) {
+            StatusCode statusCodeResponder = status.getStatusCode();
+            if(statusCodeResponder != null) {
+                StatusCode statusCode = statusCodeResponder.getStatusCode();
+                if(statusCode != null && statusCode.getValue().equals(StatusCode.AUTHN_FAILED)) {
```
Comment (sfkamath) [inline src/main/java/uk/gov/ida/eidas/bridge/helpers/responseFromEidas/ResponseHandler.java]: need to investigate this later as a BAU commit. Leaving it in for now as the suggestion breaks 5 tests

File: src/main/java/uk/gov/ida/eidas/bridge/resources/VerifyAuthnRequestResource.java (hunk 2 above)
Comment (sfkamath) [inline src/main/java/uk/gov/ida/eidas/bridge/resources/VerifyAuthnRequestResource.java]: Did go down that route and then fell back to the non functional approach. The min function  does make it readable and disambiguates the compareTo. This is now implemented as suggested

File: src/main/java/uk/gov/ida/eidas/bridge/helpers/requestToEidas/EidasAuthnRequestGenerator.java [hunk 5]
```diff
@@ -64,8 +66,18 @@ public AuthnRequest generateAuthnRequest(String authnReqeustId, String destinati
...
-        requestedAuthnContext.getAuthnContextClassRefs().add(openSamlXmlObjectFactory.createAuthnContextClassReference(LevelOfAssurance.SUBSTANTIAL.toString()));
+        RequestedAuthnContext requestedAuthnContext = openSamlXmlObjectFactory.createRequestedAuthnContext(MINIMUM_AUTHNCONTEXT);
+        String levelOfAssuranceRequested;
+        switch (lowestAuthnContext) {
+            case LEVEL_1    :   levelOfAssuranceRequested = LevelOfAssurance.LOW.toString();            break;
+            case LEVEL_2    :   levelOfAssuranceRequested = LevelOfAssurance.SUBSTANTIAL.toString();    break;
+            case LEVEL_3    :   levelOfAssuranceRequested = LevelOfAssurance.HIGH.toString();           break;
+            case LEVEL_4    :
+            default         :
+                throw new SecurityException("Unknown level of insurance from requested AuthnContext : " + lowestAuthnContext);
```
Comment (richardTowers) [inline src/main/java/uk/gov/ida/eidas/bridge/helpers/requestToEidas/EidasAuthnRequestGenerator.java]: Should be "level of assurance"

Comment (sfkamath) [inline src/main/java/uk/gov/ida/eidas/bridge/helpers/requestToEidas/EidasAuthnRequestGenerator.java]: oops, thanks. Fixed

File: src/main/java/uk/gov/ida/eidas/bridge/domain/EidasSamlResponse.java [hunk 6]
```diff
@@ -1,14 +1,33 @@
...
+        this.assertion = null;
+        this.shouldBuildAssertion = false;
     }
 
     public EidasIdentityAssertion getIdentityAssertion() {
         return assertion;
     }
+
+    public boolean isSuccess() {
+        return shouldBuildAssertion;
```
Comment (richardTowers) [inline src/main/java/uk/gov/ida/eidas/bridge/domain/EidasSamlResponse.java]: If this was `return failureStatus == null;` then you could get rid of the (now confusingly named) `shouldBuildAssertion` field.

Comment (sfkamath) [inline src/main/java/uk/gov/ida/eidas/bridge/domain/EidasSamlResponse.java]: sure, makes sense

**Pull request merged**